File: lru-cache.js
// Time Complexity : O(N)
// Space Complexity : O(N)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

// Doubly Linked List
function Node(key = null, val = null) {
    this.key = key;
    this.val = val;
    this.next = null;
    this.prev = null;
}
function removeNode(node) {
    node.next.prev = node.prev;
    node.prev.next = node.next;
}
function addToHead(node, head) {
    node.next = head.next;
    node.prev = head;
    head.next = node;
    node.next.prev = node;
}
/**
 * @param {number} capacity
 */
var LRUCache = function(capacity) {
    this.capacity = capacity;
    this.map = new Map();
    this.head = new Node(-1, -1);
    this.tail = new Node(-1, -1);
    this.head.next = this.tail;
    this.tail.prev = this.head;
};

/** 
 * @param {number} key
 * @return {number}
 */
LRUCache.prototype.get = function(key) {
    if (!this.map.has(key)) return -1;
    
    let node = this.map.get(key);
    removeNode(node);
    addToHead(node, this.head);
    return node.val;
};

/** 
 * @param {number} key 
 * @param {number} value
 * @return {void}
 */
LRUCache.prototype.put = function(key, value) {
    // When key is already in my cache
    if (this.map.has(key)) {
        let node = this.map.get(key);
        node.val = value;
        removeNode(node);
        addToHead(node, this.head);
    } else {
        if (this.map.size == this.capacity) {
            let tailPrev = this.tail.prev;
            removeNode(tailPrev);
            this.map.delete(tailPrev.key);
        }
        let newNode = new Node(key, value);
        this.map.set(key, newNode);
        addToHead(newNode, this.head);
    }
};

/** 
 * Your LRUCache object will be instantiated and called as such:
 * var obj = new LRUCache(capacity)
 * var param_1 = obj.get(key)
 * obj.put(key,value)
 */

File: LRU cache.py
class ListNode:

    def __init__(self, key, val):
        self.key = key
        self.val = val
        self.prev = None
        self.next = None


class LRUCache:

    def __init__(self, capacity):
        self.capacity = capacity
        self.hashmap = {}
        self.head = ListNode(-1, -1)
        self.tail = ListNode(-1, -1)
        self.head.next = self.tail
        self.tail.prev = self.head

    def removeNode(self, node):
        node.prev.next = node.next
        node.next.prev = node.prev

    def addTohead(self, node):
        node.prev = self.head
        node.next = self.head.next
        self.head.next = node
        node.next.prev = node

    def get(self, key) :
        # available
        if key not in self.hashmap:
            return -1

        node = self.hashmap[key]
        self.removeNode(node)
        self.addTohead(node)
        return node.val

    def put(self, key, value) :

        # present
        if key in self.hashmap:
            node = self.hashmap.get(key)
            node.val = value
            self.removeNode(node)
            self.addTohead(node)

        # not present
        else:
            if self.capacity == len(self.hashmap):
                tail_prev = self.tail.prev
                self.removeNode(tail_prev)
                del self.hashmap[tail_prev.key]
            new = ListNode(key, value)
            self.addTohead(new)
            self.hashmap[key] = new

# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)